src/compas_fab/datastructures/plan.py

```diff
@@ -0,0 +1,371 @@
+            The dependency ids to be validated.
+        planned_action_id : Hashable
+            The id of the associated planned action.  Used only in
+            the error message.  Defaults to ``None``.
+
+        Raises
+        ------
+        :class:`compas_fab.datastructures.DependencyIdException`
+        """
+        dependency_ids = set(dependency_ids)
+        if not dependency_ids.issubset(self.planned_actions):
+            invalid_ids = dependency_ids.difference(self.planned_actions)
+            raise DependencyIdException(invalid_ids, planned_action_id)
+
+    def check_all_dependency_ids(self):
+        """Checks whether the dependency ids of all the planned actions
+        are ids of planned actions in the plan.
+
+        Raises
+        ------
+        :class:`compas_fab.datastructures.DependencyIdException`
+        """
+        for pa_id, planned_action in self.planned_actions.items():
+            self.check_dependency_ids(planned_action.dependency_ids, pa_id)
+
+    def check_for_cycles(self):
+        """"Checks whether cycles exist in the dependency graph."""
+        self.check_all_dependency_ids()
+
+        def helper(cur, v, r):
+            v[cur] = True
+            r[cur] = True
+            for dep_id in self.planned_actions[cur].dependency_ids:
+                if not v[dep_id]:
+                    helper(dep_id, v, r)
+                elif r[dep_id]:
+                    raise Exception("Cycle found with action ids {}".format([pa_id for pa_id, seen in r.items() if seen]))
+            r[cur] = False
+
+        visited = {pa_id: False for pa_id in self.planned_actions}
+        rec_dict = {pa_id: False for pa_id in self.planned_actions}
+
+        for pa_id in self.planned_actions:
+            if not visited[pa_id]:
+                helper(pa_id, visited, rec_dict)
+
+    def linear_sort(self):
+        """Sorts the planned actions linearly respecting the dependency ids.
+
+        Returns
+        -------
+        :obj:`list` of :class:`compas_fab.datastructure.Action`
+        """
+        self.check_for_cycles()
+
+        def helper(s, v, cur_id):
+            v.add(cur_id)
+            action = self.planned_actions[cur_id]
+            for dep_id in action.dependency_ids:
+                if dep_id not in v:
+                    helper(s, v, dep_id)
+            s.append(action)
+
+        stack = []
+        visited = set()
+
+        for action_id in self.planned_actions:
+            if action_id not in visited:
+                helper(stack, visited, action_id)
+
+        return stack
+
+    @property
+    def data(self):
+        return dict(
+            planned_actions=list(self.planned_actions.values()),
+            id_generator=self._id_generator,
+        )
+
+    @data.setter
+    def data(self, data):
+        self.planned_actions = data['planned_actions']
+        self._id_generator = data['id_generator']
+
+    @classmethod
+    def from_data(cls, data):
+        return cls(**data)
```

Comment:
That requires that I provide default values for all of the constructor arguments, which I didn't want to do because I wanted to force the user to provide them.

Reply:
I think this will not work with default serialization when the object is nested within other things,